src/servers/Server_Zone/Script/NativeScript.cpp

```diff
@@ -0,0 +1,139 @@
+           if( script == m_questScripts.end() )
+              return nullptr;
+
+           return script->second;
+        }
+
+        BattleNpcScript* NativeScript::getBattleNpcScript( uint32_t npcId )
+        {
+           auto script = m_battleNpcScripts.find( npcId );
+           if( script == m_battleNpcScripts.end() )
+              return nullptr;
+
+           return script->second;
+        }
+
+        ZoneScript* NativeScript::getZoneScript( uint32_t zoneId )
+        {
+           auto script = m_zoneScripts.find( zoneId );
+           if( script == m_zoneScripts.end() )
+              return nullptr;
+
+           return script->second;
+        }
+
+        bool NativeScript::loadScript( std::string path )
+        {
+           auto info = m_loader.loadModule( path );
+           if( info )
+           {
+              // todo: this is shit
+              if( auto script = m_loader.getScriptObject< StatusEffectScript >( info->handle, "StatusEffectScript" ) )
+              {
+                 // todo: make this a define or something
+                 info->script = script;
+                 info->script_name = script->getName();
+                 m_statusEffectScripts[ script->getId() ] = script;
+              }
+              else if( auto script = m_loader.getScriptObject< ActionScript >( info->handle, "ActionScript" ) )
+              {
+                 info->script = script;
+                 info->script_name = script->getName();
+                 m_actionScripts[ script->getId() ] = script;
+              }
+              else if( auto script = m_loader.getScriptObject< QuestScript >( info->handle, "QuestScript" ) )
+              {
+                 info->script = script;
+                 info->script_name = script->getName();
+                 m_questScripts[ script->getId() ] = script;
+              }
+              else if( auto script = m_loader.getScriptObject< BattleNpcScript >( info->handle, "BattleNpcScript" ) )
+              {
+                 info->script = script;
+                 info->script_name = script->getName();
+                 m_battleNpcScripts[ script->getId() ] = script;
+              }
+              else if( auto script = m_loader.getScriptObject< ZoneScript >( info->handle, "ZoneScript" ) )
+              {
+                 info->script = script;
+                 info->script_name = script->getName();
+                 m_zoneScripts[ script->getId() ] = script;
+              }
+              else
+              {
+                 // unload anything which doesn't have a suitable export
+                 m_loader.unloadScript( info->handle );
+
+                 return false;
+              }
+           }
+           else
+              return false;
+
+           return true;
+        }
+
+        const std::string NativeScript::getModuleExtension()
+        {
+           return m_loader.getModuleExtension();
+        }
+
+        bool NativeScript::unloadScript( std::string name )
```

Comment:
any reason why this isn't pass by reference + const? from what I can tell you don't need mutability for name